projects/laji/src/app/+theme/kerttu/kerttu-letter-annotation/kerttu-letter-annotation.component.ts

```diff
@@ -165,7 +165,7 @@ export class KerttuLetterAnnotationComponent implements OnInit, OnDestroy {
       .pipe(
         switchMap((result) => {
           const candidate = result.candidate;
-          return this.audioService.getAudioBuffer(candidate.recording).pipe(map(() => candidate));
+          return this.audioService.getAudioBuffer(candidate.audio.url).pipe(map(() => candidate));
```

Comment:
Seems like the result from the `this.audioService.getAudioBuffer` is never used, as the `map()` just returns the `candidate`?

Reply:
It isn't used at that point but it loads the buffer to audio service cache. It makes changing of next candidate faster when the audio is already loaded.

Reply:
Hmm... could you just do the same with tap to make it more readable:

```
this.nextLetterCandidate$ = this.kerttuApi.getNextLetterCandidate(this.userService.getToken(), templateId, candidateId)
  .pipe(
    map((result) => result.candidate),
    tap(candidate => this.audioService.getAudioBuffer(candidate.audio.url).pipe(take(1)).subscribe()),
    share()
);
```

The reasoning being, that loading the audio buffer to the cache is a side effect, and the `nextLetterCandidate$` value comes from the `map` operation. The current implementation looks like the `nextLetterCandidate$` needs some value from the `switchMap` but it doesn't. This should be also faster, because you don't have to wait for the `this.audioService.getAudioBuffer` to complete before returning the `candidate`? (I didn't test the proposal, read it as a pseudo code)

Reply:
Looks good but I think it's better if these operations are in the same subscription so that they can be easily unsubscribed. So I removed the _getAudioBuffer_ call from the _nextLetterCandidate_ observable and moved it to _nextCandidateToCacheSub_. Now the names should be more descriptive.

Reply:
Ok! I don't see those changes yet, did you maybe forget to push?

Reply:
commits should be visible below ([this](https://github.com/luomus/laji/pull/80/commits/8f4aba75c50355b5465422b303191192764705c8) is the commit)